ros/src/twist_controller/twist_controller.py

```diff
@@ -6,12 +11,43 @@
 
 class Controller(object):
 
-    def __init__(self, wheel_base, steer_ratio, min_speed, max_lat_accel, max_steer_angle):
+    def __init__(self, wheel_base, steer_ratio, min_speed, max_lat_accel, max_steer_angle, max_decceleration, max_acceleration, fuel_capacity, vehicle_mass, wheel_radius):
+        self.fuel_capacity = fuel_capacity
+        self.vehicle_mass = vehicle_mass
+        self.wheel_radius = wheel_radius
+        self.current_timestep = None
+
         self.yaw_controller = YawController(
             wheel_base, steer_ratio, min_speed, max_lat_accel, max_steer_angle)
 
+        self.velocity_pid_controller = PID(
+            0.1, 0.000005, 0.05, max_decceleration, max_acceleration)
+
     def control(self, target_linear_velocity, target_angular_velocity, current_linear_velocity):
-        # TODO - implement proper throttle and brake measurements
+        # use velocity pid controller to compute new velocity
+        acceleration = self.compute_acceleration(
+            target_linear_velocity - current_linear_velocity)
+        rospy.logdebug('Current linear velocity %s, target linear velocity %s, computed acceleration %s',
+                       current_linear_velocity, target_linear_velocity, acceleration)
 
         # Return throttle, brake, steer
-        return 0., 0., self.yaw_controller.get_steering(target_linear_velocity, target_angular_velocity, current_linear_velocity)
+        return (acceleration if acceleration > 0. else 0.,
+                self.compute_brake_torque(-acceleration) if acceleration < 0. else 0.,
+                self.yaw_controller.get_steering(target_linear_velocity, target_angular_velocity, current_linear_velocity))
+
+    def reset(self):
+        self.last_timestep = None
```

Comment:
Ok for now, but forcing the accel value to 0 while self.last_timestep is None will cause the output = 0 for the first step after dbw_enabled turns ON, and then the second step output will jump back to some non-zero value.  In case of switching ON while accelerating (not likely in Carla but theoretically), this could cause a single cycle blip to zero in actual accel control value so a better way to handle resetting the D term would be to reinitialize the PID controllers self.last_error = error for this first step so that only the D term is forced to 0 but the P term can still output based on the actual error.  This is not a good explanation, let me know if it makes sense or we can discuss more later.